File: cmd/diff.go
package cmd

import (
	"fmt"
	"os"
	"os/exec"

	diagv1 "buf.build/gen/go/tldiagramcom/diagram/protocolbuffers/go/diag/v1"
	"connectrpc.com/connect"
	"github.com/mertcikla/tld-cli/client"
	"github.com/mertcikla/tld-cli/workspace"
	"github.com/spf13/cobra"
)

func newDiffCmd(wdir *string) *cobra.Command {
	c := &cobra.Command{
		Use:   "diff",
		Short: "Show differences between local workspace and server",
		RunE: func(cmd *cobra.Command, _ []string) error {
			ws, err := workspace.Load(*wdir)
			if err != nil {
				return fmt.Errorf("load workspace: %w", err)
			}

			targetOrg := ws.Config.OrgID
			if targetOrg == "" {
				return fmt.Errorf("org-id required in .tld.yaml")
			}

			// 1. Export server state to a temp directory
			tempDir, err := os.MkdirTemp("", "tld-diff-*")
			if err != nil {
				return fmt.Errorf("create temp dir: %w", err)
			}
			defer func() { _ = os.RemoveAll(tempDir) }()

			c := client.New(ws.Config.ServerURL, ws.Config.APIKey, false)
			resp, err := c.ExportOrganization(cmd.Context(), connect.NewRequest(&diagv1.ExportOrganizationRequest{
				OrgId: targetOrg,
			}))
			if err != nil {
				return fmt.Errorf("fetch server state: %w", err)
			}

			serverWS := convertExportResponse(&workspace.Workspace{Dir: tempDir, Config: ws.Config}, resp.Msg)
			if err := workspace.Save(serverWS); err != nil {
				return fmt.Errorf("save server state to temp: %w", err)
			}

			// 2. Run git diff between temp and local
			// Note: we use --no-index to diff directories not in a git repo
			// We diff FROM server TO local (so + means local addition, - means server has it but local doesn't)
			diffCmd := exec.Command("git", "diff", "--no-index", "--color=always", tempDir, *wdir)
			diffCmd.Stdout = cmd.OutOrStdout()
			diffCmd.Stderr = cmd.ErrOrStderr()

			// git diff --no-index returns 1 if differences are found, which is fine
			_ = diffCmd.Run()

			return nil
		},
	}

	return c
}

File: src/cli/diff.rs
use crate::client;
use crate::client::diagv1;
use crate::error::TldError;
use crate::output;
use crate::workspace;
use clap::Args;
use std::fs;
use std::process::Command;
use tonic::Request;

#[derive(Args, Debug, Clone)]
pub struct DiffArgs {
    /// Skip symbol verification
    #[arg(long, default_value = "false")]
    pub skip_symbols: bool,
}

#[expect(clippy::expect_used)]
pub async fn exec(_args: DiffArgs, wdir: String) -> Result<(), TldError> {
    let ws = workspace::load(&wdir)?;

    if ws.config.org_id.is_empty() {
        return Err(TldError::Generic(
            "Org ID required in .tld.yaml".to_string(),
        ));
    }

    output::print_info("Fetching server state for comparison...");

    // 1. Create temp dir
    let temp_dir = std::env::temp_dir().join(format!("tld-diff-{}", std::process::id()));
    fs::create_dir_all(&temp_dir)?;

    let mut client =
        client::new_workspace_client(&ws.config.server_url, &ws.config.api_key).await?;
    let req = Request::new(diagv1::ExportOrganizationRequest {
        org_id: ws.config.org_id.clone(),
        api_key: None,
    });

    let resp = client.export_workspace(req).await?.into_inner();

    let server_ws_dir = temp_dir
        .to_str()
        .expect("temp path should be valid utf8")
        .to_string();
    let server_ws = workspace::conversion::from_export_response(
        &server_ws_dir,
        ws.config.clone(),
        ws.ws_config.clone(),
        ws.meta.as_ref(),
        resp,
    );

    workspace::save(&server_ws)?;

    // 2. Run git diff
    output::print_info("Calculating differences (+ local addition, - server has it)...");

    let mut diff_cmd = Command::new("git");
    diff_cmd.args([
        "diff",
        "--no-index",
        "--color=always",
        &server_ws_dir,
        &wdir,
    ]);

    // We don't check exit status because git diff --no-index returns 1 if differences exist.
    let _ = diff_cmd.status();

    // 3. Cleanup
    let _ = fs::remove_dir_all(&temp_dir);

    Ok(())
}
